C++ tensor indexing: more indexing tests

yf225 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30427 C++ tensor indexing: more indexing tests**
* #30426 Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h

This PR adds the following items:
1. Translated all tensor indexing tests from Python (test/test_indexing.py) to C++.
2. Expose `at::indexing::Slice` as public API, so that users can write `a.index({Slice(), 0})` instead of `a.index({{}, 0})` to get around a Clang compile bug.

Differential Revision: [D18695899](https://our.internmc.facebook.com/intern/diff/D18695899)

Commit:
C++ tensor indexing: more indexing tests

Commit:
Update on "C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

kostmo commented:
## CircleCI build failures summary
As of commit 25a0c4d:
* **3/3** failures introduced in this PR



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=25a0c4dbe7d6bf7a50d1726630cc860267bd3b35):

### 3 failures recognized by patterns:
#### <img src='https://avatars0.githubusercontent.com/ml/7?s=12'/> binary_linux_libtorch_2_7m_cpu_gcc5_4_cxx11-abi_shared-with-deps_build
**Step:** Build
```
Nov 25 22:32:56 FAILED: test_api/CMakeFiles/test_api.dir/dataloader.cpp.o
```
#### <img src='https://avatars0.githubusercontent.com/ml/7?s=12'/> binary_linux_manywheel_2_7mu_cpu_devtoolset7_build
**Step:** Build
```
Nov 25 22:33:07 /pytorch/test/cpp/api/support.h:94:38: error: no matching function for call to ‘at::Tensor::idx_put_(<brace-enclosed initializer list>, int)’
```
#### <img src='https://avatars0.githubusercontent.com/ml/7?s=12'/> binary_linux_libtorch_2_7m_cpu_devtoolset7_shared-with-deps_build
**Step:** Build
```
Nov 25 22:33:14 /pytorch/test/cpp/api/support.h:94:38: error: no matching function for call to ‘at::Tensor::idx_put_(<brace-enclosed initializer list>, int)’
```


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

Commit:
Update on "[WIP] C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "[WIP] C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "[WIP] C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "[WIP] C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "[WIP] C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "[WIP] C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "[WIP] C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "[WIP] C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "[WIP] C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "[WIP] C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "[WIP] C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "[WIP] C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "[WIP] C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "[WIP] C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "[WIP] C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "[WIP] C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "[WIP] C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "[WIP] C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "[WIP] C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "[WIP] C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "[WIP] C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "[WIP] C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

yf225 commented:
TODO: To fix the following error on OSX build:
```
Feb 25 10:36:32 ../test/cpp/api/tensor_indexing.cpp:694:46: error: call to member function 'index' is ambiguous
Feb 25 10:36:32   assert_tensor_equal(a.index({0, "..."}), a.index({0, {}}));
Feb 25 10:36:32                                            ~~^~~~~
Feb 25 10:36:32 aten/src/ATen/core/TensorBody.h:400:10: note: candidate function
Feb 25 10:36:32   Tensor index(std::initializer_list<at::indexing::TensorIndex> indices) const;
Feb 25 10:36:32          ^
Feb 25 10:36:32 aten/src/ATen/core/TensorBody.h:399:10: note: candidate function
Feb 25 10:36:32   Tensor index(ArrayRef<at::indexing::TensorIndex> indices) const;
Feb 25 10:36:32          ^
Feb 25 10:36:32 aten/src/ATen/core/TensorMethods.h:1386:23: note: candidate function
Feb 25 10:36:32 inline Tensor Tensor::index(TensorList indices) const {
Feb 25 10:36:32                       ^
Feb 25 10:36:32 ../test/cpp/api/tensor_indexing.cpp:695:46: error: call to member function 'index' is ambiguous
Feb 25 10:36:32   assert_tensor_equal(a.index({"...", 0}), a.index({{}, 0}));
Feb 25 10:36:32                                            ~~^~~~~
Feb 25 10:36:32 aten/src/ATen/core/TensorBody.h:400:10: note: candidate function
Feb 25 10:36:32   Tensor index(std::initializer_list<at::indexing::TensorIndex> indices) const;
Feb 25 10:36:32          ^
Feb 25 10:36:32 aten/src/ATen/core/TensorBody.h:399:10: note: candidate function
Feb 25 10:36:32   Tensor index(ArrayRef<at::indexing::TensorIndex> indices) const;
Feb 25 10:36:32          ^
Feb 25 10:36:32 aten/src/ATen/core/TensorMethods.h:1386:23: note: candidate function
Feb 25 10:36:32 inline Tensor Tensor::index(TensorList indices) const {
Feb 25 10:36:32                       ^
Feb 25 10:36:32 2 warnings and 2 errors generated.
```
I need to rename the `Tensor::index(TensorList indices)` function to `Tensor::advanced_index(TensorList indices)`, and `Tensor::index_put_(TensorList indices, ...)` to `Tensor::advanced_index_put_(TensorList indices, ...)`

Commit:
Update on "C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head

Commit:
Update on "C++ tensor indexing: more indexing tests"

C++ tensor indexing: more indexing tests

gh-metadata: pytorch pytorch 30427 gh/yf225/89/head
